networks/decoder_net.py

```diff
@@ -0,0 +1,130 @@
+                kernel_size=1,
+                bias=False,
+                padding=0
+            )
+        self.leaky_relu = nn.LeakyReLU()
+        self.upsample = nn.UpsamplingNearest2d(scale_factor=2)
+        self.sigmoid = nn.Sigmoid()
+
+    def forward(self, x):
+        """Apply 2x up-sampling, followed by two convolutional layers with leaky relu. A sigmoid
+        activation is applied at the end.
+
+        TODO: Should we use batch normalization? It is often common in residual blocks.
+        TODO: Here we apply a convolutional layer to the input up-sampled tensor if its number
+            of channels does not match the convolutional layer channels. Is this the correct way?
+
+        Args:
+            x (torch.Tensor): Input image of shape (N, C, H, W) where N is the batch size and C
+                is the number of in_channels.
+
+        Returns:
+            A torch.Tensor with the upsampled images, of shape (N, n_filters, H, W).
+        """
+        x = self.upsample(x)
+        residual = self.dim_match_conv(x) if self.channels != self.n_filters else x
+        x = self.leaky_relu(self.conv1(x))
+        x = self.leaky_relu(self.conv2(x))
+        x = self.sigmoid(x + residual)
+        return x
+
+
+class DecoderNet(nn.Module):
+    """The Decoder network, that takes a latent encoding of shape (in_channels, H, W)
+    and produces the output image by applying 3 ResidualBlock modules and a final 1x1 convolution.
+    Each residual block up-scales the image by 2, and the convolution produces the desired number
+    of output channels, thus the output shape is (out_channels, H*2^3, W*2^3).
+    """
+
+    def __init__(self, in_channels, out_channels, n_filters):
+        """Create the decoder network.
+
+        Args:
+            in_channels (int): Number of input encodings channels.
+            out_channels (int): Number output image channels (1 for grayscale, 3 for RGB).
+            n_filters (int): Number of filters in the intermediate convolutional layers.
+        """
+        super().__init__()
+        self.res1 = ResidualBlock(in_channels=in_channels, n_filters=n_filters)
+        self.res2 = ResidualBlock(in_channels=n_filters, n_filters=n_filters)
+        self.res3 = ResidualBlock(in_channels=n_filters, n_filters=n_filters)
+        self.out = nn.Conv2d(
+            in_channels=n_filters,
+            out_channels=out_channels,
+            kernel_size=1,
+            padding=0,
+        )
+        self.sigmoid = nn.Sigmoid()
+
+    def forward(self, x):
+        """Apply the three residual blocks and the final convolutional layer.
+
+        Args:
+            x (torch.Tensor): Tensor of shape (N, in_channels, H, W) where N is the batch size.
+
+        Returns:
+            Tensor of shape (out_channels, H * 2^3, W * 2^3) with the reconstructed image.
+        """
+        x = self.res1(x)
+        x = self.res2(x)
+        x = self.res3(x)
+        x = self.sigmoid(self.out(x))
```

Comment:
They dont say anything about sigmoid activation in the lasc Conv layer... Did you get it from karras et al 2018?

Reply:
No I didn't. You're right, maybe this is wrong. It feels weird that an NN that outputs an image doesn't have a sigmoid activation at the end since we want all outputs to be in [0, 1].

Reply:
I removed it for now. We will check this later. I'm adding it to #9